File: 784-letter-case-permutation/784-letter-case-permutation.cpp
class Solution {
public:
    vector<string> letterCasePermutation(string s) {
        vector<string> res;
        backtrack(s,0,res);
        return res;
    }
     void backtrack(string &s, int i, vector<string> &res) {
        if (i == s.size()) {
            res.push_back(s);
            return;
        }
        backtrack(s, i + 1, res);
        if (isalpha(s[i])) {
            // toggle case
            s[i] ^= (1 << 5);
            backtrack(s, i + 1, res);
        }
    }
};

File: 784-letter-case-permutation/784-letter-case-permutation.java
class Solution {
    public List<String> letterCasePermutation(String s) {
        if (s == null) 
            return new LinkedList<>();
        List<String> res = new LinkedList<>();
        helper(s.toCharArray(), res, 0);
        return res;
    } 
    public void helper(char[] chs, List<String> res, int pos) {
        if (pos == chs.length) {
            res.add(new String(chs));
            return;
        }
        if (chs[pos] >= '0' && chs[pos] <= '9') {
            helper(chs, res, pos + 1);
            return;
        }
        chs[pos] = Character.toLowerCase(chs[pos]);
        helper(chs, res, pos + 1);
        
        chs[pos] = Character.toUpperCase(chs[pos]);
        helper(chs, res, pos + 1);
    }
}